### ISSUE TITLE
WebAudio backend doesn't stream audio, but downloads entire piece, decodes entire piece and only then starts playing [infamous audio lag problem]

### ISSUE THREAD

**Issue by ghost** [rpgtkoolmv/corescript#27]
If you check js/rpg_core/WebAudio.js, you'll find this function which is called every time a sound is created & triggered for playback (including new music pieces):
```js
WebAudio.prototype.initialize = function(url) {
    if (!WebAudio._initialized) {
        WebAudio.initialize();
    }
    this.clear();
    this._load(url);
    this._url = url;
};
```
You can see it calls `_load`, which triggers an Ajax request:
```js
WebAudio.prototype._load = function(url) {
    if (WebAudio._context) {
        var xhr = new XMLHttpRequest();
        if(Decrypter.hasEncryptedAudio) url = Decrypter.extToEncryptExt(url);
        xhr.open('GET', url);
....
```
In the receive callback of the Ajax request, _onXhrLoad simply decodes the received audio after the Ajax request completed in its entirety (full download required first):
```js
WebAudio.prototype._onXhrLoad = function(xhr) {
    var array = xhr.response;
    if(Decrypter.hasEncryptedAudio) array = Decrypter.decryptArrayBuffer(array);
    this._readLoopComments(new Uint8Array(array));
   WebAudio._context.decodeAudioData(array, function(buffer) {
...
```

.. and only *after* that, WebAudio.prototype._onLoad is called which finally informs all load listeners.
As you can see in the play function, the load listeners are used to start playing the actual sound, so before reaching this stage no audio can be heard:

```js
WebAudio.prototype.play = function(loop, offset) {
    if (this.isReady()) {
        offset = offset || 0;
        this._startPlaying(loop, offset);
    } else if (WebAudio._context) {
        this._autoPlay = true;
        this.addLoadListener(function() {
            if (this._autoPlay) {
                this.play(loop, offset);
            }
        }.bind(this));
    }
};
```
In conclusion, this means any sound or music piece is first download as a whole, then decoded as a whole and only then it starts playing. This process can take many actual seconds, I observed delays up to 6-7 seconds for a song to start playing even in a local(!) game where no download and only decoding is required. In HTML5-delivered web games, the problem is obviously potentially a lot worse.

This is the infamous audio lag problem that plagues RPG Maker MV for all things audio. Therefore, as a conclusion this code really should be changed so that:

1. decoding starts before the download is completed
2. playback starts before decoding is completed

Then, all audio lag problems with sounds and music should be solved, except for cases with a really sluggy internet connection in HTML5 games. Those can only ever work nicely if everything is preloaded completely, which is by the way also what this plugin workaround hack does: https://forums.rpgmakerweb.com/index.php?threads/webaudio-cache-experimental.51422/ but obviously this is bad for regular desktop games since it will need more memory and increase loading time by a lot, so this isn't a good solution.
```

**Comment by ghost** [comment]
Scope of the problem: WebAudio appears to be the default audio backend for anything but sometimes mobile (Android/iOS), so this affects all desktop RPG Maker MV games which is I would argue the most important release platform. I don't know how the Html5Audio backend fares in comparison, so this would need a separate investigation.

**Comment by ghost** [comment]
We know about the problem.(see road map)

And We found the problem is hard to solve in current browser feature.
If you have idea for implement streaming, please PR.

BTW, I don't know why the original script doesn't use MediaElementAudioSourceNode.
does anyone know that?

**Comment by ghost** [comment]
I asked original author why it doesn't.

**Comment by ghost** [comment]
The answer returns.
The API is not usable when development starts.

But for now, It may be usable (needs investigation)
So we consider about using the API.

This fix will be done at 1.6.
But PR is welcome anytime!!
